planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,562 @@
+  }
+
+  private void removeShortStubEdges() {
+    PriorityQueue<Edge> toRemove = new PriorityQueue<>(Comparator.comparingDouble(Edge::length));
+    for (var node : output) {
+      for (var edge : node.getEdges()) {
+        if (isShortStubEdge(edge)) {
+          toRemove.offer(edge);
+        }
+      }
+    }
+    while (!toRemove.isEmpty()) {
+      var edge = toRemove.poll();
+      if (edge.removed) {
+        continue;
+      }
+      edge.remove();
+      if (degreeTwoMerge(edge.from) instanceof Edge merged && isShortStubEdge(merged)) {
+        toRemove.offer(merged);
+      }
+      if (edge.from.getEdges().size() == 1) {
+        var other = edge.from.getEdges().getFirst();
+        if (isShortStubEdge(other)) {
+          toRemove.offer(other);
+        }
+      }
+      if (edge.from != edge.to) {
+        if (degreeTwoMerge(edge.to) instanceof Edge merged && isShortStubEdge(merged)) {
+          toRemove.offer(merged);
+        }
+        if (edge.to.getEdges().size() == 1) {
+          var other = edge.to.getEdges().getFirst();
+          if (isShortStubEdge(other)) {
+            toRemove.offer(other);
+          }
+        }
+      }
+    }
+  }
+
+  private boolean isShortStubEdge(Edge edge) {
+    return edge != null && !edge.removed && edge.length < stubMinLength &&
+      (edge.from.getEdges().size() == 1 || edge.to.getEdges().size() == 1 || edge.from == edge.to);
+  }
+
+  private void removeShortEdges() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < minLength) {
+          edge.remove();
+        }
+      }
+    }
+  }
+
+  private void simplify() {
+    for (var node : output) {
+      for (var edge : node.getEdges()) {
+        if (edge.main) {
+          edge.simplify();
+        }
+      }
+    }
+  }
+
+  private void removeDuplicatedEdges() {
+    for (var node : output) {
+      List<Edge> toRemove = new ArrayList<>();
+      for (var i = 0; i < node.getEdges().size(); ++i) {
+        Edge a = node.getEdges().get(i);
+        for (var j = i + 1; j < node.getEdges().size(); ++j) {
+          Edge b = node.getEdges().get(j);
+          if (a.coordinates.equals(b.coordinates)) {
+            toRemove.add(b);
+          }
+        }
+      }
+      for (var edge : toRemove) {
+        edge.remove();
+      }
+    }
+  }
+
+  /**
+   * Processes the added geometries and returns the merged linestrings.
+   * <p>
+   * Can be called more than once.
+   */
+  public List<LineString> getMergedLineStrings() {
+    output.clear();
+    List<List<Coordinate>> edges = nodeLines(input);
+    buildNodes(edges);
+
+    degreeTwoMerge();
+
+    if (loopMinLength > 0.0) {
+      breakLoops();
+      degreeTwoMerge();
+    }
+
+    if (stubMinLength > 0.0) {
+      removeShortStubEdges();
+      // removeShortStubEdges does degreeTwoMerge internally
+    }
+
+    if (tolerance >= 0.0) {
+      simplify();
+      removeDuplicatedEdges();
```

Comment:
Why do we end up with duplicate edges here? Is it possible to remove this duplicate edge removal step?